File: lc328.java
class Solution {
    public ListNode oddEvenList(ListNode head) {
   
        if(head==null || head.next==null || head.next.next==null)
            return head;
        ListNode temp = head;
        ListNode oddStart = null;
        ListNode oddEnd = null;
        ListNode evenStart = null;
        ListNode evenEnd = null;
        int id = 1;
        while(temp!=null)
        {
            if(id%2==1) //odd
            {
                if(oddStart==null)
                {
                    oddStart = temp;
                    oddEnd = temp;
                }
                else
                {
                    oddEnd.next = temp;
                    oddEnd = oddEnd.next;
                }   
                temp = temp.next;                        
                oddEnd.next = null;
            }
            else
            {
                if(evenStart==null)
                {
                    evenStart = temp;
                    evenEnd = temp;
                }
                else
                {
                    evenEnd.next = temp;
                    evenEnd = evenEnd.next;
                }          
                temp = temp.next;    
                evenEnd.next = null;
            }
            id++;
        }
        oddEnd.next = evenStart;
        return oddStart;
    }
}

File: python/lc328.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head:
            oddStart = head
            oddEnd = head
            head = head.next
            oddEnd.next = None
            if head:        
                evenStart = head
                evenEnd = head        
                head = head.next
                evenEnd.next = None
            else:
                return oddStart
        else:
            return head
        odd = True
        while head:
            if odd:
                oddEnd.next = head
                oddEnd = oddEnd.next
                head = head.next
                oddEnd.next = None
                odd = False
            else:
                evenEnd.next = head
                evenEnd = evenEnd.next
                head = head.next
                evenEnd.next = None
                odd = True
            
        oddEnd.next = evenStart
        return oddStart



